File: packages/core/src/contracts/chart-presentation.ts
import type {ChartNumberFormat} from './chart-theme'

/** How an axis maps values to positions. `'auto'` picks from the data. */
export type ChartAxisScale = 'auto' | 'category' | 'linear' | 'log' | 'time'

/** Which side of the plot an axis is drawn on. */
export type ChartAxisPosition = 'end' | 'start'

/** Pins an axis extent. Omit either end to keep it computed from the data. */
export type ChartAxisDomain = {
  max?: number
  min?: number
  /**
   * Headroom kept beyond the data, as a fraction of its extent — `0.08` leaves 8%
   * clear at each end. Applies only to an end that is computed, so it can be
   * combined with a single pinned one.
   *
   * Without it a line runs into the top and bottom of its own plot, which reads as
   * clipped rather than as the highest and lowest reading.
   */
  padding?: number
}

/**
 * Full control of one axis. `axis={{x: false}}` is still the short way to hide
 * one; reach for this when you need a pinned domain, a log scale or exact ticks.
 */
export type ChartAxisOptions = {
  domain?: ChartAxisDomain
  format?: ChartNumberFormat
  grid?: boolean
  gridDash?: readonly number[]
  label?: string
  labelRotation?: number
  position?: ChartAxisPosition
  reversed?: boolean
  scale?: ChartAxisScale
  /** A hint. The renderer still picks readable ticks around it. */
  tickCount?: number
  /** Exact ticks, when the reader is looking for specific values. */
  tickValues?: readonly (number | string)[]
  /** Draws the axis at all. Hiding one takes its grid with it — the rules belong to the scale. */
  visible?: boolean
}

/** Inset between the plot area and its container. */
export type ChartPlotPadding = {
  bottom?: number
  left?: number
  right?: number
  top?: number
}

/** Styles the plot area only, not the surface around it. */
export type ChartPlotStyle = {
  backgroundColor?: string
  borderColor?: string
  borderRadius?: number
  borderWidth?: number
  /** Clips marks to the plot area. Wanted whenever a domain is pinned. */
  clip?: boolean
  padding?: number | ChartPlotPadding
}

/** The shape drawn at a data point. */
export type ChartSymbol = 'circle' | 'diamond' | 'none' | 'square' | 'triangle'

/** Overrides for one series, keyed by `ChartSeries.id`. */
export type ChartSeriesStyle = {
  color?: string
  fillOpacity?: number
  opacity?: number
  strokeDash?: readonly number[]
  strokeWidth?: number
  symbol?: ChartSymbol
  symbolSize?: number
}

/** The curve an animation follows. */
export type ChartAnimationEasing = 'ease-in' | 'ease-in-out' | 'ease-out' | 'linear' | 'spring'

/** Timing for the entrance and for later data changes. */
export type ChartAnimation = {
  /**
   * How long the entrance waits before it starts, in ms. Worth a beat when the chart arrives
   * on a pushed screen: the first part of a trace drawn under a navigation transition is
   * never seen.
   */
  delay?: number
  duration?: number
  easing?: ChartAnimationEasing
  enabled?: boolean
  /** The entrance. Turn it off for a chart that re-renders on every keystroke. */
  initial?: boolean
  /** The transition when data changes under a mounted chart. */
  updates?: boolean
}

/** Which crosshair lines follow the pointer. */
export type ChartCrosshairMode = 'both' | 'none' | 'x' | 'y'

/** What a hover picks up: the whole axis slice, the nearest mark, or one series. */
export type ChartHoverMode = 'axis' | 'nearest' | 'none' | 'series'

/** How many marks can be selected at once. */
export type ChartSelectionMode = 'multiple' | 'none' | 'single'

/** Everything about how a chart responds to a pointer or a finger. */
export type ChartInteraction = {
  crosshair?: ChartCrosshairMode
  /**
   * How long the marks take to step back to `dimOpacity` when a reading starts, and to come back up
   * when it ends, in ms. Default 0, which is instant — give it a beat when the dimming is the whole
   * gesture's feedback, so a finger landing reads as the lights coming down rather than as a cut.
   */
  dimDuration?: number
  /** How far the other marks fade while one is hovered. */
  dimOpacity?: number
  /** Native only. The web has no equivalent. */
  haptics?: boolean
  /**
   * How far the read mark is blended towards `highlightColor`, 0–1. Default 1. Below that
   * the mark's own colour shows through, so a red candle lights up red rather than white.
   */
  highlightBlend?: number
  /**
   * A colour the mark under the pointer is blended towards, so the read one reads as lit
   * rather than merely as the one that was not dimmed.
   */
  highlightColor?: string
  /** How far a hovered mark grows. */
  highlightScale?: number
  hover?: ChartHoverMode
  pan?: boolean
  /**
   * Reads a span rather than a mark: two fingers on the plot report the marks under each of
   * them, and everything between. iOS and Android only — a pointer has no second finger.
   *
   * The span arrives as `range` on the interaction event, beside the single `index` one
   * finger reports, so the same `onInteraction` serves both and the readout decides which
   * of the two it is showing.
   */
  range?: boolean
  selection?: ChartSelectionMode
  tooltip?: boolean
  zoom?: boolean
}

/** A position on an axis: a number, or a category name. */
export type ChartCoordinate = number | string

/**
 * Where in a category's band a rule sits. A band is a width, not a line, so a rule placed on
 * one has to pick: `'center'` runs through the mark, `'start'` and `'end'` run along the band's
 * own edges — which is where a rule that means "up to here" belongs, since the mark it names is
 * inside the span rather than the end of it.
 *
 * Only means anything on a category axis. Default `'center'`.
 */
export type ChartBandAlign = 'center' | 'end' | 'start'

/** A reference line across the plot: a target, a threshold, a launch date. */
export type ChartLineAnnotation = {
  /** Where in a category's band the rule sits. Ignored on a numeric axis. Default `'center'`. */
  align?: ChartBandAlign
  axis: 'x' | 'y'
  color?: string
  /**
   * Dash and gap lengths, in points. Omit for a solid rule — a dash of `[6, 3]` reads as
   * a boundary, `[1, 4]` as a reference the eye can cross.
   */
  dash?: readonly number[]
  id: string
  label?: string
  type: 'line'
  value: ChartCoordinate
  /** How thick the rule is drawn, in points. Default 1. */
  width?: number
}

/** A shaded span: a quarter, an incident window, a tolerance band. */
export type ChartRangeAnnotation = {
  axis: 'x' | 'y'
  color?: string
  end: ChartCoordinate
  id: string
  label?: string
  opacity?: number
  start: ChartCoordinate
  type: 'range'
}

/** A single marked point, placed by coordinate rather than by data. */
export type ChartPointAnnotation = {
  color?: string
  id: string
  label?: string
  symbol?: ChartSymbol
  type: 'point'
  x: ChartCoordinate
  y: number
}

/** Free text placed on the plot. Omit `x`/`y` to let the renderer place it. */
export type ChartTextAnnotation = {
  color?: string
  id: string
  text: string
  type: 'text'
  x?: ChartCoordinate
  y?: number
}

/** Anything drawn on top of the data. Discriminated by `type`. */
export type ChartAnnotation = ChartLineAnnotation | ChartPointAnnotation | ChartRangeAnnotation | ChartTextAnnotation

/**
 * Where a scrub is in its lifetime. `'began'` and `'ended'` bracket one gesture,
 * so a readout outside the plot knows when to return to its resting value.
 */
export type ChartInteractionPhase = 'began' | 'changed' | 'ended' | 'layout'

/** Where one annotation ended up, in the chart's own coordinate space. */
export type ChartAnnotationGeometry = {
  id: string
  x: number
  y: number
}

/**
 * The plot's box and its annotations' positions, in the chart's own coordinate space.
 * Reported on the `'layout'` phase so an app can place its own views over the chart —
 * its own badge on an annotation, its own card at a reading — instead of taking the ones
 * the chart draws.
 */
export type ChartGeometry = {
  annotations: readonly ChartAnnotationGeometry[]
  plot: {height: number; width: number; x: number; y: number}
}

/**
 * The span two fingers are reading, in data order: `startIndex` is always the lower of the
 * two, however the fingers are placed, and both ends are inclusive.
 *
 * The edge positions come with it because the chart is the one that knows where a category
 * landed — a card centred over the span needs them, and an app cannot compute them from the
 * plot's width without also knowing the axis padding and the bar inset.
 */
export type ChartInteractionRange = {
  endCategory?: string
  endIndex: number
  /** Where the mark at `endIndex` sits, in the chart's own coordinate space. */
  endX?: number
  startCategory?: string
  startIndex: number
  /** Where the mark at `startIndex` sits, in the chart's own coordinate space. */
  startX?: number
}

/** What `onInteraction` receives. Fields are filled in as the chart form allows. */
export type ChartInteractionEvent = {
  category?: string
  /** Where the plot and its annotations sit, on the `'layout'` phase. */
  geometry?: ChartGeometry
  /** Position of the selected mark in the chart's own data order. */
  index?: number
  /** Pointer position in the chart's own coordinate space. */
  nativeX?: number
  nativeY?: number
  phase?: ChartInteractionPhase
  /** The span under two fingers, when `interaction.range` is on and both are down. */
  range?: ChartInteractionRange
  seriesId?: string
  /** Unix seconds, on the time-based forms. */
  timestamp?: number
  value?: number
  x?: number
  y?: number
}

File: packages/zyplot/android/src/main/java/com/hzblj/zyplot/core/presentation/ChartPresentation.kt
package com.hzblj.zyplot.core.presentation

import java.util.Locale
import org.json.JSONObject

data class AxisDomain(
  val minimum: Double?,
  val maximum: Double?,
)

/**
 * Mirrors `ChartNumberFormat`. Chart props are serialized across the bridge, so a
 * caller cannot hand us a formatter function — it describes one instead.
 *
 * The fallback locale is the constant `en-US` rather than the device default, so a
 * chart formats identically to its web counterpart unless the caller names a
 * locale explicitly.
 */
data class NumberFormat(
  val decimals: Int,
  val locale: String?,
  val prefix: String,
  val suffix: String,
) {
  fun format(value: Double): String {
    val resolved = locale?.takeIf { it.isNotEmpty() }
      ?.let(Locale::forLanguageTag)
      ?: Locale.US
    return prefix + String.format(resolved, "%,.${decimals}f", value) + suffix
  }

  companion object {
    fun from(json: JSONObject?): NumberFormat = NumberFormat(
      decimals = (json?.nullableDouble("decimals")?.toInt() ?: 0).coerceIn(0, 20),
      locale = json?.nullableString("locale"),
      prefix = json?.nullableString("prefix") ?: "",
      suffix = json?.nullableString("suffix") ?: "",
    )
  }
}

data class AxisOptions(
  val domain: AxisDomain,
  val format: JSONObject?,
  val grid: Boolean,
  val gridDash: List<Float>,
  val label: String?,
  val labelOverflow: String,
  val labelRotation: Float,
  val position: String,
  val reversed: Boolean,
  val scale: String,
  val tickCount: Int,
  val visible: Boolean?,
) {
  companion object {
    fun from(json: JSONObject?): AxisOptions {
      val domain = json?.optJSONObject("domain")
      return AxisOptions(
        domain = AxisDomain(
          minimum = domain?.nullableDouble("min"),
          maximum = domain?.nullableDouble("max"),
        ),
        format = json?.optJSONObject("format"),
        grid = json?.optBoolean("grid", true) ?: true,
        gridDash = json?.optJSONArray("gridDash")?.let { array ->
          (0 until array.length()).map { array.optDouble(it).toFloat() }
        } ?: emptyList(),
        label = json?.nullableString("label"),
        labelOverflow = json?.nullableString("labelOverflow") ?: "ellipsis",
        labelRotation = json?.optDouble("labelRotation", 0.0)?.toFloat() ?: 0f,
        position = json?.optString("position", "start") ?: "start",
        reversed = json?.optBoolean("reversed", false) ?: false,
        scale = json?.optString("scale", "auto") ?: "auto",
        tickCount = (json?.optInt("tickCount", 4) ?: 4).coerceAtLeast(1),
        visible = if (json?.has("visible") == true) json.optBoolean("visible") else null,
      )
    }
  }
}

data class PlotPadding(
  val bottom: Float,
  val left: Float,
  val right: Float,
  val top: Float,
) {
  companion object {
    fun from(value: Any?): PlotPadding = when (value) {
      is Number -> PlotPadding(
        bottom = value.toFloat(),
        left = value.toFloat(),
        right = value.toFloat(),
        top = value.toFloat(),
      )
      is JSONObject -> PlotPadding(
        bottom = value.optDouble("bottom", 0.0).toFloat(),
        left = value.optDouble("left", 0.0).toFloat(),
        right = value.optDouble("right", 0.0).toFloat(),
        top = value.optDouble("top", 0.0).toFloat(),
      )
      else -> PlotPadding(0f, 0f, 0f, 0f)
    }
  }
}

data class PlotStyle(
  val backgroundColor: String?,
  val borderColor: String?,
  val borderRadius: Float,
  val borderWidth: Float,
  val clip: Boolean,
  val padding: PlotPadding,
) {
  companion object {
    fun from(json: JSONObject?): PlotStyle = PlotStyle(
      backgroundColor = json?.nullableString("backgroundColor"),
      borderColor = json?.nullableString("borderColor"),
      borderRadius = json?.optDouble("borderRadius", 0.0)?.toFloat() ?: 0f,
      borderWidth = json?.optDouble("borderWidth", 0.0)?.toFloat() ?: 0f,
      clip = json?.optBoolean("clip", true) ?: true,
      padding = PlotPadding.from(json?.opt("padding")),
    )
  }
}

data class SeriesStyle(
  val color: String?,
  val fillOpacity: Float,
  val opacity: Float,
  val strokeDash: List<Float>,
  val strokeWidth: Float?,
  val symbol: String,
  val symbolSize: Float,
) {
  companion object {
    fun from(json: JSONObject): SeriesStyle = SeriesStyle(
      color = json.nullableString("color"),
      fillOpacity = json.optDouble("fillOpacity", 0.16).toFloat(),
      opacity = json.optDouble("opacity", 1.0).toFloat(),
      strokeDash = json.optJSONArray("strokeDash")?.let { array ->
        (0 until array.length()).map { array.optDouble(it).toFloat() }
      } ?: emptyList(),
      strokeWidth = json.nullableDouble("strokeWidth")?.toFloat(),
      symbol = json.optString("symbol", "none"),
      symbolSize = json.optDouble("symbolSize", 7.0).toFloat(),
    )
  }
}

data class AnimationOptions(
  val delayMillis: Int,
  val durationMillis: Int,
  val easing: String,
  val enabled: Boolean,
  val initial: Boolean,
  val updates: Boolean,
) {
  companion object {
    fun from(json: JSONObject?): AnimationOptions = AnimationOptions(
      delayMillis = json?.optInt("delay", 0) ?: 0,
      durationMillis = json?.optInt("duration", 320) ?: 320,
      easing = json?.optString("easing", "ease-out") ?: "ease-out",
      enabled = json?.optBoolean("enabled", true) ?: true,
      initial = json?.optBoolean("initial", true) ?: true,
      updates = json?.optBoolean("updates", true) ?: true,
    )
  }
}

data class InteractionOptions(
  val crosshair: String,
  val dimOpacity: Float,
  val haptics: Boolean,
  val highlightScale: Float,
  val hover: String,
  val pan: Boolean,
  val selection: String,
  val tooltip: Boolean,
  val zoom: Boolean,
) {
  val isEnabled: Boolean
    get() = hover != "none" || crosshair != "none" || selection != "none" || tooltip

  companion object {
    fun from(json: JSONObject?): InteractionOptions = InteractionOptions(
      crosshair = json?.optString("crosshair", "none") ?: "none",
      dimOpacity = json?.optDouble("dimOpacity", 0.25)?.toFloat() ?: 0.25f,
      haptics = json?.optBoolean("haptics", false) ?: false,
      highlightScale = json?.optDouble("highlightScale", 1.0)?.toFloat() ?: 1f,
      hover = json?.optString("hover", "none") ?: "none",
      pan = json?.optBoolean("pan", false) ?: false,
      selection = json?.optString("selection", "none") ?: "none",
      tooltip = json?.optBoolean("tooltip", true) ?: true,
      zoom = json?.optBoolean("zoom", false) ?: false,
    )
  }
}

data class ChartAnnotation(
  val axis: String?,
  val color: String?,
  val dash: List<Float>,
  val end: Any?,
  val id: String,
  val label: String?,
  val opacity: Float,
  val start: Any?,
  val text: String?,
  val type: String,
  val value: Any?,
  val x: Any?,
  val y: Double?,
) {
  companion object {
    fun from(json: JSONObject): ChartAnnotation = ChartAnnotation(
      axis = json.nullableString("axis"),
      color = json.nullableString("color"),
      dash = json.optJSONArray("dash")?.let { array ->
        (0 until array.length()).map { array.optDouble(it).toFloat() }
      } ?: emptyList(),
      end = json.opt("end").takeUnless { it == JSONObject.NULL },
      id = json.optString("id"),
      label = json.nullableString("label"),
      opacity = json.optDouble("opacity", 0.12).toFloat(),
      start = json.opt("start").takeUnless { it == JSONObject.NULL },
      text = json.nullableString("text"),
      type = json.optString("type"),
      value = json.opt("value").takeUnless { it == JSONObject.NULL },
      x = json.opt("x").takeUnless { it == JSONObject.NULL },
      y = json.nullableDouble("y"),
    )
  }
}

internal fun JSONObject.nullableString(name: String): String? =
  optString(name).takeIf { it.isNotEmpty() && it != "null" }

internal fun JSONObject.nullableDouble(name: String): Double? =
  if (has(name) && !isNull(name)) optDouble(name) else null
